File: app/src/main/java/com/chess/cryptobot/util/CoinInfo.java
package com.chess.cryptobot.util;

import com.chess.cryptobot.exceptions.MarketException;
import com.chess.cryptobot.exceptions.SyncServiceException;
import com.chess.cryptobot.market.Market;
import com.chess.cryptobot.model.response.CurrenciesResponse;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.chess.cryptobot.market.Market.BITTREX_MARKET;
import static com.chess.cryptobot.market.Market.LIVECOIN_MARKET;

public class CoinInfo {
    private final Map<String, Map<String, Boolean>> statuses = new HashMap<>();
    private final Map<String, Map<String, Double>> fees = new HashMap<>();

    public CoinInfo(List<Market> markets) throws MarketException {
        for (Market market : markets) {
            List<CurrenciesResponse> currencies = market.getCurrencies();
            Map<String, Boolean> statuses = new HashMap<>();
            Map<String, Double> fees = new HashMap<>();
            currencies.forEach(currency -> {
                String currencyName = currency.getCurrencyName();
                statuses.put(currencyName, currency.isActive());
                fees.put(currencyName, currency.getFee());
            });
            this.statuses.put(market.getMarketName(), statuses);
            this.fees.put(market.getMarketName(), fees);
        }
    }

    public boolean checkCoinStatus(String coinName) {
        Map<String, Boolean> bittrexStatuses = statuses.get(BITTREX_MARKET);
        if (bittrexStatuses == null) return false;
        Map<String, Boolean> livecoinStatuses = statuses.get(LIVECOIN_MARKET);
        if (livecoinStatuses == null) return false;

        Boolean bittrexStatus = bittrexStatuses.get(coinName);
        Boolean livecoinStatus = livecoinStatuses.get(coinName);

        if (bittrexStatus == null || livecoinStatus == null) return false;
        return (bittrexStatus && livecoinStatus);
    }

    public Double getFee(String marketName, String coinName) throws SyncServiceException {
        Map<String, Double> fees = this.fees.get(marketName);
        if (fees == null) {
            throw new SyncServiceException("Can't get fees");
        }

        Double fee = fees.get(coinName);
        if (fee == null) {
            throw new SyncServiceException("Can't get fee from " + marketName);
        }

        return fee;
    }

}

File: app/src/main/java/com/chess/cryptobot/util/CoinInfo.kt
package com.chess.cryptobot.util

import com.chess.cryptobot.exceptions.SyncServiceException
import com.chess.cryptobot.market.Market
import java.util.*

class CoinInfo(markets: List<Market?>) {
    private val statuses: MutableMap<String, Map<String?, Boolean?>> = HashMap()
    private val fees: MutableMap<String, Map<String?, Double?>> = HashMap()
    fun checkCoinStatus(coinName: String): Boolean {
        val bittrexStatuses = statuses[Market.BITTREX_MARKET] ?: return true
        val binanceStatuses = statuses[Market.BINANCE_MARKET] ?: return true
        val livecoinStatuses = statuses[Market.LIVECOIN_MARKET] ?: return true
        val bittrexStatus = bittrexStatuses[coinName]
        val binanceStatus = binanceStatuses[coinName]
        val livecoinStatus = livecoinStatuses[coinName]
        return if (bittrexStatus == null || binanceStatus == null || livecoinStatus == null) true
        else bittrexStatus && binanceStatus && livecoinStatus
    }

    @Throws(SyncServiceException::class)
    fun getFee(marketName: String, coinName: String): Double {
        val fees = fees[marketName]
                ?: throw SyncServiceException("Can't get fees")
        return fees[coinName] ?: throw SyncServiceException("Can't get fee from $marketName")
    }

    init {
        markets.parallelStream().forEach {
            it ?: return@forEach
            val currencies = it.getCurrencies()
            synchronized(this) {
                val fees: MutableMap<String?, Double?> = HashMap()
                val statuses: MutableMap<String?, Boolean?> = HashMap()
                for (currency in currencies) {
                    val currencyName = currency.currencyName
                    statuses[currencyName] = currency.isActive
                    fees[currencyName] = currency.fee
                }

                this.statuses[it.getMarketName()] = statuses
                this.fees[it.getMarketName()] = fees
            }
        }
    }
}